File: frontend/src/api/analysis.ts
import { requestJson } from './http'
import { openEventSource } from './http'
import type { StockDailySnapshot, StockInstrument } from './stocks'
import { buildQueryString } from './query'

export type FactorWeightItemResponse = {
  factor_key: string
  factor_label: string
  weight: number
  direction: string
  evidence: string[]
  reason: string
}

export type AnalysisEventResponse = {
  event_id: string
  scope: string
  title: string
  published_at: string | null
  source: string
  macro_topic: string | null
  event_type: string | null
  event_tags: string[] | null
  sentiment_label: string | null
  sentiment_score: number | null
  anchor_trade_date: string | null
  window_return_pct: number | null
  window_volatility: number | null
  abnormal_volume_ratio: number | null
  correlation_score: number | null
  confidence: string | null
  link_status: string | null
}

export type AnalysisReportResponse = {
  id?: string | null
  status: string
  summary: string
  risk_points: string[]
  factor_breakdown: FactorWeightItemResponse[]
  generated_at: string
  anchor_event_id?: string | null
  anchor_event_title?: string | null
  structured_sources?: Array<{ provider?: string; count?: number }>
  trigger_source?: 'manual' | 'watchlist_daily'
  used_web_search?: boolean
  web_search_status?: 'used' | 'disabled' | 'unsupported'
  session_id?: string | null
  started_at?: string | null
  completed_at?: string | null
  content_format?: 'markdown'
  web_sources?: Array<{
    title?: string
    url?: string
    source?: string
    published_at?: string | null
    snippet?: string | null
    domain?: string | null
    metadata_status?: 'enriched' | 'domain_inferred' | 'unavailable'
  }>
}

export type StockAnalysisSummaryResponse = {
  ts_code: string
  instrument: StockInstrument | null
  latest_snapshot: StockDailySnapshot | null
  status: 'ready' | 'partial' | 'pending'
  generated_at: string | null
  topic: string | null
  event_context_status?: 'direct' | 'topic_fallback' | 'none'
  event_context_message?: string | null
  published_from: string | null
  published_to: string | null
  event_count: number
  events: AnalysisEventResponse[]
  report: AnalysisReportResponse | null
}

export type AnalysisReportArchiveListResponse = {
  ts_code: string
  items: AnalysisReportResponse[]
}

export type AnalysisSessionCreateResponse = {
  session_id: string | null
  report_id: string | null
  status: string
  reused: boolean
  cached: boolean
}

export type AnalysisSessionStatusEvent = {
  session_id: string
  status: string
}

export type AnalysisSessionDeltaEvent = {
  session_id: string
  delta: string
  content: string
}

export type AnalysisSessionCompletedEvent = {
  session_id: string
  report_id: string | null
  status: string
}

export type AnalysisSessionErrorEvent = {
  session_id?: string
  detail: string
}

export const analysisApi = {
  async getStockAnalysisSummary(
    tsCode: string,
    options?: { topic?: string | null; eventId?: string | null },
  ) {
    const query = buildQueryString({
      topic: options?.topic || undefined,
      event_id: options?.eventId || undefined,
    })
    return requestJson<StockAnalysisSummaryResponse>(
      `/api/analysis/stocks/${encodeURIComponent(tsCode)}/summary${query}`,
    )
  },

  async getStockAnalysisReports(
    tsCode: string,
    limit = 10,
    options?: { topic?: string | null; eventId?: string | null },
  ) {
    const query = buildQueryString({
      limit,
      topic: options?.topic || undefined,
      event_id: options?.eventId || undefined,
    })
    return requestJson<AnalysisReportArchiveListResponse>(
      `/api/analysis/stocks/${encodeURIComponent(tsCode)}/reports${query}`,
    )
  },

  async createAnalysisSession(
    tsCode: string,
    payload: {
      topic?: string | null
      event_id?: string | null
      force_refresh?: boolean
      use_web_search?: boolean
      trigger_source?: 'manual' | 'watchlist_daily'
    },
  ) {
    return requestJson<AnalysisSessionCreateResponse>(
      `/api/analysis/stocks/${encodeURIComponent(tsCode)}/sessions`,
      {
        method: 'POST',
        body: payload,
      },
    )
  },

  openAnalysisSessionEvents(
    sessionId: string,
    handlers: {
      onStatus?: (payload: AnalysisSessionStatusEvent) => void
      onReused?: (payload: AnalysisSessionStatusEvent) => void
      onDelta?: (payload: AnalysisSessionDeltaEvent) => void
      onCompleted?: (payload: AnalysisSessionCompletedEvent) => void
      onError?: (payload: AnalysisSessionErrorEvent) => void
    },
    options?: { reused?: boolean },
  ) {
    const query = options?.reused ? '?reused=true' : ''
    return openEventSource(`/api/analysis/sessions/${encodeURIComponent(sessionId)}/events${query}`, {
      status: (payload) => handlers.onStatus?.(payload as AnalysisSessionStatusEvent),
      reused: (payload) => handlers.onReused?.(payload as AnalysisSessionStatusEvent),
      delta: (payload) => handlers.onDelta?.(payload as AnalysisSessionDeltaEvent),
      completed: (payload) => handlers.onCompleted?.(payload as AnalysisSessionCompletedEvent),
      error: (payload) => handlers.onError?.(payload as AnalysisSessionErrorEvent),
    })
  },
}

File: backend/app/schemas/analysis.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

from app.schemas.stocks import StockDailySnapshotResponse, StockInstrumentResponse


class FactorWeightItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    factor_key: str
    factor_label: str
    weight: float
    direction: str
    evidence: list[str]
    reason: str


class AnalysisEventLinkResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    event_id: str
    scope: str
    title: str
    published_at: datetime | None
    source: str
    macro_topic: str | None
    event_type: str | None
    event_tags: list[str]
    sentiment_label: str | None
    sentiment_score: float | None
    anchor_trade_date: date | None
    window_return_pct: float | None
    window_volatility: float | None
    abnormal_volume_ratio: float | None
    correlation_score: float | None
    confidence: str | None
    link_status: str | None


class AnalysisReportResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str | None = None
    status: Literal["ready", "partial", "pending"]
    summary: str
    risk_points: list[str]
    factor_breakdown: list[FactorWeightItemResponse]
    generated_at: datetime | None
    topic: str | None
    published_from: datetime | None
    published_to: datetime | None
    trigger_source: str = "manual"
    # used_web_search 标识是否实际调用检索能力，避免与开关混淆。
    used_web_search: bool = False
    web_search_status: Literal["used", "disabled", "unsupported"] = "disabled"
    session_id: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    content_format: Literal["markdown"] = "markdown"
    anchor_event_id: str | None = None
    anchor_event_title: str | None = None
    # structured_sources/web_sources 用于前端来源展示，允许为空数组。
    structured_sources: list[dict[str, object]] = []
    web_sources: list[dict[str, object]] = []


class AnalysisReportArchiveItemResponse(AnalysisReportResponse):
    pass


class AnalysisReportArchiveListResponse(BaseModel):
    ts_code: str
    items: list[AnalysisReportArchiveItemResponse]


class AnalysisSessionCreateRequest(BaseModel):
    topic: str | None = None
    event_id: str | None = None
    # force_refresh 为真时绕过缓存，强制重新生成。
    force_refresh: bool = False
    use_web_search: bool = False
    trigger_source: Literal["manual", "watchlist_daily"] = "manual"


class AnalysisSessionCreateResponse(BaseModel):
    session_id: str | None
    report_id: str | None = None
    status: str
    reused: bool = False
    cached: bool = False


class StockAnalysisSummaryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    ts_code: str
    instrument: StockInstrumentResponse | None
    latest_snapshot: StockDailySnapshotResponse | None
    status: Literal["ready", "partial", "pending"]
    generated_at: datetime | None
    topic: str | None
    # event_context_status 表示事件上下文获取口径，用于解释摘要来源。
    event_context_status: Literal["direct", "topic_fallback", "none"] = "none"
    event_context_message: str | None = None
    published_from: datetime | None
    published_to: datetime | None
    event_count: int
    events: list[AnalysisEventLinkResponse]
    report: AnalysisReportResponse | None
